libs/librrgraph/src/base/rr_graph_builder.cpp

```diff
@@ -37,12 +37,36 @@ vtr::vector<RRNodeId, std::vector<short>>& RRGraphBuilder::node_ptc_storage() {
     return node_ptc_nums_;
 }
 
-void RRGraphBuilder::add_node_to_all_locs(RRNodeId node) {
+static short calculate_node_ptc(t_graph_type graph_type, short node_ptc_base, Direction node_dir, int offset){
+    if(graph_type == t_graph_type::GRAPH_UNIDIR_TILEABLE){
+        if(node_dir != Direction::NONE){ //only chanX and chanY have different ptc number compared to non-tileable graph
+            //Incrementing wires should have their ptc number increased by 2, otherwise decremented by 2 in each step
+            int fac = (node_dir == Direction::INC) ? 2 : -2;
+            return node_ptc_base + fac*offset;
+        }
+        else{ //other rr node types (sink,source,IPIN,OPIN)
+            return node_ptc_base;
+        }
+    }
+    else{
+        return node_ptc_base;
+    }
+
+    //unreachable statement
+    return node_ptc_base;
+}
+
+void RRGraphBuilder::add_node_to_all_locs(RRNodeId node, t_graph_type graph_type) {
     t_rr_type node_type = node_storage_.node_type(node);
-    short node_ptc_num = node_storage_.node_ptc_num(node);
     short node_layer = node_storage_.node_layer(node);
+    short node_ptc_base = node_storage_.node_ptc_num(node);
+    Direction node_dir = (node_type == CHANX || node_type == CHANY) ? node_storage_.node_direction(node): Direction::NONE;
+
+    int offset = 0;
     for (int ix = node_storage_.node_xlow(node); ix <= node_storage_.node_xhigh(node); ix++) {
         for (int iy = node_storage_.node_ylow(node); iy <= node_storage_.node_yhigh(node); iy++) {
+            short node_ptc_num = calculate_node_ptc(graph_type, node_ptc_base, node_dir , offset);
```

Comment:
This may be dangerous, potentially leading to some issues in OpenFPGA, e.g., netlist error. I need to double check.

Reply:
Can you list the scenarios where this might fail? 
Certainly, there is a scope for improvement, but for now, it generated exact same netlist for a couple of architectures I tested.

Reply:
```
grep "add_node_to_all_locs" */*.cpp ../../vtr-verilog-to-routing/vpr/src/*/*.cpp ../../vtr-verilog-to-routing/libs/librrgraph/src/*/*.cpp
../../vtr-verilog-to-routing/vpr/src/route/clock_network_builders.cpp:    /* TODO: Will replace these codes with an API add_node_to_all_locs() of RRGraphBuilder */
../../vtr-verilog-to-routing/vpr/src/route/clock_network_builders.cpp:    /* TODO: Will replace these codes with an API add_node_to_all_locs() of RRGraphBuilder */
../../vtr-verilog-to-routing/libs/librrgraph/src/base/rr_graph_builder.cpp:void RRGraphBuilder::add_node_to_all_locs(RRNodeId node) {
```

Actually, the API ``add_node_to_all_locs()`` is not used by tileable rr graph builder. So the changes here may not work on tileable rr graph builder. Can you clarify your motivation? If it is just for the native rr graph builder of VPR?

Reply:
The add_node_to_all_locs() belongs to the process that tries to read rr_graph back in not while generating it. This function works fine while you try to read an rrgraph generated without tileable option. But if you have tileable = true in the architecture and generate an rrgraph, you can't use --read_rr_graph anymore with the generated rr graph, you will get a node mismatch error since some data structure in that process don't know about ptc assignment in tileable true. Hence, the update goes to reading the rr graph functions not the one that are generating it. 
I am not sure why you think it wouldn't work on tileable rr graph builder; since we have generated many netlists and GSB files with tileable=true and checked the rr graph and read it back in with the same option. If you clarify your question, I can probably give you a better answer.

Reply:
@saaramahmoudi I see the pain point: you have no track information when reading a rr_graph file because the track information is built on-fly during rr_graph builder functions. 
I think it is better not to do the calculation in the I/O as it may bring more restrictions in future. My opinion is that such calculation should be handled by the engineer which crafts rr graph files. I/O functions just need to do read-and-assign or get-and-write jobs. 
I believe a better way is to extend the writer of rr graph files which can output a list of track numbers. And we just read back these numbers through reader.

I believe your current changes may work on some specific testcases. However, since these testcases are not open to me, I can judge based on the code changes that I see. If you can more provide some testcases or try some existing testcases on OpenFPGA. It will be a better proof.

Reply:
@tangxifan Your suggestion sounds good to me. However, if we change the write functions, we may ended up not supporting existing rr-graphs for both vpr baseline and openfpga (tileable = true) options. 
I almost test all of circuit and architectures available in openfpga_flow with new function. Plus, VTR CI tests have some example of reading rr graph as well (like symbiflow test case) and those are also passing, meaning that baseline is working as expected.  
If you have any testcase in mind, I would be happy to test it out or you can think of test that It won't work, please let me know

Reply:
@saaramahmoudi The changes on writer function should be fairly straightforward. 

![image](https://github.com/verilog-to-routing/vtr-verilog-to-routing/assets/11499826/cba05822-28cb-48ae-a142-a20ed5a3cfab)

- I/O function can accept a rr_node with multiple ptc number. No checks will be enforced for now (unless we see some demands). For instance, ``ptc="1,2,3,4"``
- In regular routing resource graph, there is only one ptc number. So this will not change anything. When parsing, you can assign the same ptc number to all the node coordinates. For example, ``ptc="1"``
- When parsing tileable routing resource graph, there could be a number of ptc number. They will be all parsed and assigned, just like what you did in the ``add_node_to_all_locs()``.

I agree that your codes change may work in some architectures and netlists. However, it may force restrictions when building routing resource graphs. When we design the ``RRGraphView`` and ``RRGraphBuilder``, we made the APIs as doer functions, which means they just do as builder functions want. These APIs should be general purpose, while builder functions do the calculation and customization. Similarly, I/O functions are also doer functions. It should be user who decide how their routing resource graphs look like.

I have suggested some tests in my reply to @ganeshgore . Feel free to try.

Reply:
I agree with @tangxifan here, it is better to change the rrgraph writer so that the generated rrgraph doesn't have any special properties/interpretation and it is entirely compatible with the rrgraph generated by the VPR.